### ISSUE TITLE
[BUG] Deadlock with nested loops

### ISSUE THREAD

**Issue by lczech** [bshoshany/thread-pool#95]
**Describe the bug**

The program deadlocks when tasks in the pool submit tasks themselves to the pool again. This is useful in libraries where different parts can be parallelized, e.g., an inner function can be called in a context where itself can make use of multiple threads, but it might also be called in a context that itself is already running in parallel. Hence, this should ideally work, for example, as in OpenMP, where (as far as I am aware) nested parallel loops just re-use existing threads without deadlocking, e.g., by running the nested code in the same thread.

**Minimal working example**

Using thread-pool 3.3.0

```cpp
#include <iostream>
#include <string>
#include "BS_thread_pool.hpp"
using namespace BS;

// Global variables, for simplicity here.
synced_stream sync_out;
thread_pool tp_pool(2);

void sleep(size_t sec = 1)
{
    // We add a sleep in each task, so that multiple tasks get submitted first,
    // before the pool starts running them, so that they have a chance to submit their nested tasks.
    std::this_thread::sleep_for(std::chrono::milliseconds(1000 * sec));
}

void work(size_t i)
{
    // Do some work, by submitting nested tasks.
    sync_out.println("  sub B ", i);
    auto result = tp_pool.submit([=](){
        sync_out.println("  beg B ", i);
        sleep();
        sync_out.println("  end B ", i);
    });
    sync_out.println("  sup B ", i);

    // Deadlock here, because the task never gets executed,
    // as the pool is still executing the tasks submitted from main.
    result.get();
    sync_out.println("  fin B ", i);
}

int main()
{
    sync_out.println("---");

    // Add some outer tasks that themselves submitted work tasks.
    std::vector<std::future<void>> tasks;
    for (size_t i = 0; i < 4; ++i) {
        sync_out.println("sub A ", i);
        tasks.emplace_back( tp_pool.submit([=](){
            sync_out.println("beg A ", i);
            sleep();
            work(i);
            sync_out.println("end A ", i);
        }));
        sync_out.println("sup A ", i);
    }

    // Wait for everything.
    for( auto& task : tasks ) {
        task.get();
    }
    tp_pool.wait_for_tasks();
    sync_out.println("---");

    return 0;
}
```

**Behavior**

Prints

```
---
sub A 0
sup A 0
sub A 1
sup A 1
sub A 2
sup A 2
sub A 3
sup A 3
beg A 0
beg A 1
  sub B 1
  sub B 0
  sup B 1
  sup B 0
```

then deadlocks, because the pool is executing the A-level tasks, and waiting for them to finish, but they never can, as the B-level tasks are waiting at the end of the queue, and would only get executed once the A's are done.

**Additional information**

Potential solution: Run nested tasks in the same thread that their submitting task is run in.

PS: You do you, but I am perplexed by the contribution policy of this repository - not accepting PRs, and only pushing final releases makes it difficult and not appealing to be a contributor to this repository. Off thread, I'd be interested in the rational for this - what is the advantage over a more open policy? "Testing things locally" could easily be changed to some CI (GitHub Actions for example) that could check PRs as well ;-)

**Comment by bshoshany** [comment]
Hi @lczech, thanks for opening this issue. Sorry it took me so long to reply. I'm currently on hiatus from developing this package, since I'm too busy teaching. However, I plan to release a new version in the summer.

I ran your code and was indeed able to reproduce the deadlock. However, I'm not sure how to solve this.

First, when you call `tp_pool.submit()` from within a task in the pool, as far as I know, there is no way for the pool to know that this was called from a task and not from the main thread.

Also, I'm not sure if running the nested task on the same thread as the submitting task is possible using the current thread pool algorithm. If the thread is running the submitting task, then the only way to return control to the worker (in order to run another task) is to end the task. If the task is still running, and trying to read a future from nested tasks, then the worker will never regain control.

I guess one way to do this is to replace `submit()` with something like `submit_nested()` which will take as an argument the ID of the current thread and know to run that task on **another** thread. However, this will only resolve the deadlock if there is at least one other thread available, which is not guaranteed. (If I have 4 threads and each one is trying to run a nested task, then nothing will happen.)

This library is very low-level, it doesn't "hold the user's hand", so perhaps the best way to solve this is simply for the user to be aware that nested tasks will cause a deadlock, and know either not to nest tasks in the first place, or to have a separate pool for the nested tasks.

So my plan is to simply add this as an example to the documentation, with an attached warning. I am closing this issue since I have added this to my TODO list with high priority, and it will be implemented in the next release.

However, I am open to other suggestions!

As for the contribution policy: I think pull requests and GitHub Actions are great tools for large collaborations where many people work on the same codebase. But this is just a small package that I developed on my own, and I like doing things my own way. I put it on GitHub so that other people can use it, and make suggestions via issues and PRs. Many suggestions have been extremely helpful, but I still prefer to implement any changes locally on my own.

I like to do extensive testing on several different devices (including an HPC cluster), as well as performance tests, and this isn't really something GitHub Actions can do at the moment, since I can't exactly ask it to run the code on an 80-thread CPU. GitHub Actions is a good way to make sure a PR doesn't break the code, but it's not convenient for actual development. I like running the code on my own system, changing things here and there, debugging, writing short test programs to test specific features, and so on.

I hope you can understand, and I do greatly appreciate your contribution!

**Comment by lczech** [comment]
I've implemented a solution that checks for the thread id, and if it's the same as the one that the thread pool was started from, then instead of queuing the task, it's immediately run within the enqueue function. That is, threads that submit their own nested tasks have a different id, so then their nested task is run immediately (within that very thread) instead of enqueuing it. That avoids the deadlock.

Obviously that would fail if two pools send tasks to each other, or similar scenarios. But with one pool, it seems to solve the issue, and I have not come up with scenarios where this would be suboptimal.
